src/network/network_gui.cpp

```diff
@@ -1564,333 +1575,735 @@ NetworkCompanyInfo *GetLobbyCompanyInfo(CompanyID company)
+				/* Always ask confirmation for all admin actions. */
+				ShowQuery(
+					STR_NETWORK_CLIENT_LIST_ASK_CAPTION,
+					text,
+					this,
+					callback
+				);
+			} break;
+
+			default:
+				NOT_REACHED();
 		}
-		return true;
+
+		this->SetDirty();
 	}
 
-	void UpdateWidgetSize(int widget, Dimension *size, const Dimension &padding, Dimension *fill, Dimension *resize) override
+	void OnQueryTextFinished(char *str) override
 	{
-		if (widget != WID_CL_PANEL) return;
+		if (str == nullptr) return;
 
-		this->server_client_width = std::max(GetStringBoundingBox(STR_NETWORK_SERVER).width, GetStringBoundingBox(STR_NETWORK_CLIENT).width) + WD_FRAMERECT_RIGHT;
-		this->icon_size = GetSpriteSize(SPR_COMPANY_ICON);
-		this->line_height = std::max(this->icon_size.height + 2U, (uint)FONT_HEIGHT_NORMAL);
+		switch (this->query_widget) {
+			default: NOT_REACHED();
 
-		uint width = 100; // Default width
-		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-			width = std::max(width, GetStringBoundingBox(ci->client_name).width);
-		}
+			case WID_CL_SERVER_NAME_EDIT: {
+				if (!_network_server) break;
+
+				uint index;
+				GetSettingFromName("network.server_name", &index);
+				SetSettingValue(index, StrEmpty(str) ? "My Server" : str);
+				this->InvalidateData();
+				break;
+			}
 
-		size->width = WD_FRAMERECT_LEFT + this->server_client_width + this->icon_size.width + WD_FRAMERECT_LEFT + width + WD_FRAMERECT_RIGHT;
+			case WID_CL_CLIENT_NAME_EDIT: {
+				uint index;
+				GetSettingFromName("network.client_name", &index);
+				SetSettingValue(index, StrEmpty(str) ? "Player" : str);
+				this->InvalidateData();
+				break;
+			}
+
+			case WID_CL_COMPANY_JOIN:
+				NetworkClientRequestMove(this->join_company, str);
+				break;
+		}
 	}
 
-	void OnPaint() override
+	uint DrawButtons(uint x, uint y, const std::vector<std::unique_ptr<ButtonCommon>> *buttons) const
 	{
-		/* Check if we need to reset the height */
-		if (!this->CheckClientListHeight()) return;
+		for (auto &button : *buttons) {
+			bool rtl = _current_text_dir == TD_RTL;
 
-		this->DrawWidgets();
+			uint left = rtl ? x : x - button->width;
+			uint right = rtl ? x + button->width : x;
+
+			int offset = std::max(0, ((int)(this->line_height + 1) - (int)button->height) / 2);
+
+			DrawFrameRect(left, y + offset, right, y + offset + button->height, button->colour, FR_NONE);
+			DrawSprite(button->sprite, PAL_NONE, left + ScaleGUITrad(WD_FRAMERECT_LEFT), y + offset + ScaleGUITrad(WD_FRAMERECT_TOP));
+			if (button->disabled) {
+				GfxFillRect(left + 1, y + offset + 1, right - 1, y + offset + button->height - 1, _colour_gradient[button->colour & 0xF][2], FILLRECT_CHECKER);
+			}
+
+			uint width = button->width + WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT;
+			x += rtl ? width : -width;
+		}
+
+		return x;
 	}
 
-	void DrawWidget(const Rect &r, int widget) const override
+	uint DrawCompany(CompanyID company_id, uint left, uint right, uint top, uint count) const
```

Comment:
Maybe rename `count` to `line` as that seems to more accurately convey the meaning of the parameter; as far as I can see it is not the total amount of companies.